samples/net/http_server/src/main.c

```diff
@@ -0,0 +1,718 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <errno.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/logging/log_ctrl.h>
+#include <zephyr/net/net_ip.h>
+#include <zephyr/net/socket.h>
+#include <zephyr/sys/reboot.h>
+
+#include <zephyr/net/net_config.h>
+#include <zephyr/net/net_mgmt.h>
+#include <zephyr/net/net_event.h>
+#include <zephyr/net/conn_mgr_monitor.h>
+#include <zephyr/net/conn_mgr_connectivity.h>
+
+#include <zephyr/net/http/client.h>
+#include <zephyr/net/http/parser.h>
+
+#include <dk_buttons_and_leds.h>
+
+#include "credentials_provision.h"
+
+LOG_MODULE_REGISTER(http_server, CONFIG_HTTP_SERVER_SAMPLE_LOG_LEVEL);
+
+#define SERVER_PORT			CONFIG_HTTP_SERVER_SAMPLE_PORT
+#define MAX_CLIENT_QUEUE		CONFIG_HTTP_SERVER_SAMPLE_CLIENTS_MAX
+#define STACK_SIZE			CONFIG_HTTP_SERVER_SAMPLE_STACK_SIZE
+#define THREAD_PRIORITY			K_PRIO_COOP(CONFIG_NUM_COOP_PRIORITIES - 1)
+
+/* Macro used to subscribe to specific Zephyr NET management events. */
+#define L4_EVENT_MASK (NET_EVENT_L4_CONNECTED | NET_EVENT_L4_DISCONNECTED)
+#define CONN_LAYER_EVENT_MASK (NET_EVENT_CONN_IF_FATAL_ERROR)
+
+/* Macro called upon a fatal error, reboots the device. */
+#define FATAL_ERROR()								\
+	LOG_ERR("Fatal error!%s", IS_ENABLED(CONFIG_RESET_ON_FATAL_ERROR) ?	\
+				  " Rebooting the device" : "");		\
+	LOG_PANIC();								\
+	IF_ENABLED(CONFIG_REBOOT, (sys_reboot(0)))
+
+/* Zephyr NET management event callback structures. */
+static struct net_mgmt_event_callback l4_cb;
+static struct net_mgmt_event_callback conn_cb;
+
+struct http_req {
+	struct http_parser parser;
+	int socket;
+	bool received_all;
+	enum http_method method;
+	const char *url;
+	size_t url_len;
+	const char *body;
+	size_t body_len;
+};
+
+/* Forward declarations */
+static void process_tcp4(void);
+static void process_tcp6(void);
+
+/* Keep track of the current LED states. 0 = LED OFF, 1 = LED ON.
+ * Index 0 corresponds to LED1, index 1 to LED2.
+ */
+static uint8_t led_states[2];
+
+/* HTTP responses for demonstration */
+#define response_200 "HTTP/1.1 200 OK\r\n"
+#define response_400 "HTTP/1.1 400 Bad Request\r\n\r\n"
+#define response_403 "HTTP/1.1 403 Forbidden\r\n\r\n"
+#define response_404 "HTTP/1.1 404 Not Found\r\n\r\n"
+#define response_405 "HTTP/1.1 405 Method Not Allowed\r\n\r\n"
+#define response_500 "HTTP/1.1 500 Internal Server Error\r\n\r\n"
```

Comment:
Should we make the macros all caps?
```suggestion
#define RESPONSE_200 "HTTP/1.1 200 OK\r\n"
#define RESPONSE_400 "HTTP/1.1 400 Bad Request\r\n\r\n"
#define RESPONSE_403 "HTTP/1.1 403 Forbidden\r\n\r\n"
#define RESPONSE_404 "HTTP/1.1 404 Not Found\r\n\r\n"
#define RESPONSE_405 "HTTP/1.1 405 Method Not Allowed\r\n\r\n"
#define RESPONSE_500 "HTTP/1.1 500 Internal Server Error\r\n\r\n"
```

Reply:
added